src/translator/request.h

```diff
@@ -27,88 +17,119 @@
+  /// marian-decoder's maxi-batching-mechanism. histories returns embed line
+  /// numbers which are used in OutputCollector
   size_t lineNumberBegin_;
 
-  // Multiple translation-workers can concurrently access the same Request. The
-  // following atomic atomically operates on the variable holding sentences
-  // remaining to be translated.
+  /// Multiple translation-workers can concurrently access the same Request. The
+  /// following atomic atomically operates on the variable holding sentences
+  /// remaining to be translated.
   std::atomic<int> counter_;
 
-  // source_ holds the source string to be translated. segments_ hold the
-  // sentences generated from source_ in vector<Words>. sourceRanges_ are
-  // string_views of the text corresponding to these words, pointing to
-  // sequences in source_. histories_ is a buffer which eventually stores the
-  // translations of each segment in the corresponding index.
+  /// source_ holds the source string to be translated.
   std::string source_;
+  /// segments_ hold the sentences generated from source_ in vector<Words>.
   Segments segments_;
+  /// sourceRanges_ are string_views of the text corresponding to these words,
+  /// pointing to sequences in source_.
   SentenceRanges sourceRanges_;
+  /// histories_ is a buffer which eventually stores the  translations of each
+  /// segment in the corresponding index.
   std::vector<Ptr<History>> histories_;
 
-  // Members above are moved into newly constructed Response on completion
-  // of translation of all segments. The promise below is set to this Response
-  // value. future to this promise is made available to the user through
-  // Service.
-  std::promise<Response> response_;
-
-  // Constructing Response requires the vocabs_ used to generate Request.
+  /// Constructing Response requires the vocabs_ used to generate Request.
   std::vector<Ptr<Vocab const>> *vocabs_;
+
+  /// Stores callback function to be called on complete.
+  std::function<void()> onComplete_{
+      []() {} // no-op by default.
+  };
+  std::promise<Response> response_;
 };
 
+/// RequestSentence is a tuple of (index, Ptr<Request>). This provides the
+/// batching mechanism access to the segment within the request. The backref to
+/// Request allows event triggering the barrier upon completion of the last
+/// sentence by a worker.
+
 class RequestSentence {
-  // A RequestSentence provides a view to a sentence within a Request. Existence
-  // of this class allows the sentences and associated information to be kept
-  // within Request.
 
 public:
   RequestSentence(size_t, Ptr<Request>);
   size_t numTokens() const;
 
-  // lineNumber in Request, used for matching marian-decoder. SentenceTuple
-  // requires lineNumber to be set for Corpus based batches.
+  /// lineNumber in Request, used for matching marian-decoder. SentenceTuple
+  /// requires lineNumber to be set for Corpus based batches.
```

Comment:
lineNumber is expected to go away, until I can match marian-decoder and replace it I'm stuck with lineNumberBegin and the remaining. This is becoming an increasing point of pain during review, that I'm thinking of writing my own OutputCollector and discard histories_ altogether.

Reply:
OutputCollector was one of the first things that I chucked out in my REST server implementation. I kept histories because they store all the information that any output-producing front-end may need. I nevertheless would prefer sequenceNumber (or seqNo) over lineNumber, because that's what it basically is.